### ISSUE TITLE
Double splat operator with empty hash works differently in JRuby 9.3 and MRI 2.6

### ISSUE THREAD

**Issue by comandeo** [jruby/jruby#7184]
**Environment Information**
`jruby 9.3.3.0 (2.6.8) 2022-01-19 b26de1f5c5 OpenJDK 64-Bit Server VM 17.0.3+7-Ubuntu-0ubuntu0.20.04.1 on 17.0.3+7-Ubuntu-0ubuntu0.20.04.1 +jit [linux-x86_64] `
and
`jruby 9.3.4.0 (2.6.8) 2022-03-23 eff48c1ebf OpenJDK 64-Bit Server VM 17.0.3+7-Ubuntu-0ubuntu0.20.04.1 on 17.0.3+7-Ubuntu-0ubuntu0.20.04.1 +jit [x86_64-linux]`

**Expected Behavior**
Below is a simple test that demonstrates different behavior of double splat operator with an empty hash for Ruby versions 2.6 and below and 2.7 and above.

```ruby
class Example
  def bar(key, opt1: false, opt2: true, opt3: nil)
    #
  end
end

RSpec.describe Example do
  it do
    options = {}

    if RUBY_VERSION >= '2.7'
      expect(subject).to receive(:bar).with(:key, **options)
    else
      expect(subject).to receive(:bar).with(:key, options)
    end

    subject.bar(:key, **options)
  end
end
```
This test passes on MRI 2.6.3, 2.7.4, and 3.1.0. It also passes on JRuby 9.2.20.1.

**Actual Behavior**

The same test fails on JRuby 9.3.3.0 and 9.3.4.0 with the following error:
```
Failures:

  1) Example is expected to receive bar(:key, {}) 1 time
     Failure/Error: subject.bar(:key, **options)

       #<Example:0x260d48f5> received :bar with unexpected arguments
         expected: (:key, {})
              got: (:key)
       Diff:
       @@ -1 +1 @@
       -[:key, {}]
       +[:key]
```
`RUBY_VERSION` constant for JRuby 9.3.3.0 and 9.3.4.0 is set to `"2.6.8"`, so the test sets expectation according to MRI 2.6 behavior; however, it seems that the actual behavior matches MRI 2.7.

Here is the link to a test in Mongoid where we encountered this issue - https://github.com/mongodb/mongoid/blob/27122e92e492b25729e0fc657f67753bf6bd8fbb/spec/mongoid/errors/mongoid_error_spec.rb#L12

**Comment by enebo** [comment]
This is not fun.  This {} is not getting passed anymore from #6564.  I have not looked at rspec and how it implements argument capturing but I did find a pretty simple reproduction:

```ruby
def bar(*args)
  p args
end

options = {}
bar(:key, **options)
```

In #6564 I sort of knew this was an incompatibility but did not think of how common this may be (Ruby 2.7+ will give the same behavior 9.3.4.0 does).  Luckily I just went through a newer redesign of Ruby keywords for upcoming 9.4.  The downside is that we eliminated the argument before we execute the call.  So the receiving method has no idea there was a keyword argument supplied.

**Comment by enebo** [comment]
If I remove the **{} stripping logic from this then the reported example is fixed (which I expected but it is always helpful to prove it).

**Comment by enebo** [comment]
Very interesting.  I removed the conditional logic but left the rest of the fix for #6564 in place.  Everything passes except using an explicit ```**{}``` when being passed to a method with  ```*args```.  This could also get fixed with more work but it seems so unlikely that I will not bother (until someone reports this).  Note: ```**something_empty``` is fine.  I honestly find 2.6 behavior here to be wrong.  They are both empty hashes so why does one not pass?  Seems like it is just a strange optimization which appears to be behavior.  Do people really write ```**{}``` in actual code and not language test suites?

The reason the removal of conditional logic works for #6564 and this issue is that the code inserts a MERGE_KWARGS instr to the kwrest value.  Part of this merge is to dup the hash and then have the hash mark itself a kw_rest.  This used to be how upcoming 9.4 did things (uses the hash to mark what it is) but it ended up being broken in the presence of threads.  In the case of Ruby 2.6 and the fact this is the only place it is used makes it fine since it will be dup'd.
